src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -11,80 +11,515 @@
+    ucc_rank_t                          dst_rank;
+    void                               *src_addr;
+    void                               *dst_addr;
+    ucs_status_ptr_t                    request;
+    size_t                              red_idx;
+    size_t                              put_offset;
+    int                                 window;
+    int                                 put_idx;
+    ucp_ep_h                            ep;
+
+    ucc_assert(host_team_size > 0);
+
+    if (allgather_task != NULL) {
+        ucc_tl_ucp_allreduce_sliding_window_key_exchange_progress(coll_task);
+        return;
+    }
+
+    if (*reduce_task != NULL) {
+        // We've previously started a reduction on the accbuf that hasn't yet
+        // completed.
+        ucc_tl_ucp_allreduce_sliding_window_test_reduction(task);
+
+        if (*reduce_task != NULL) {
+            return;
+        }
+    }
+
+    if (pipe->count_serviced < pipe->my_count) {
+        if ((pipe->count_received < pipe->my_count) &&
+            (pipe->done_get < host_team_size) && (pipe->avail_buffs > 0) &&
+            (accbuf->state != REDUCED && accbuf->state != SENDING)) {
+            remaining_elems = pipe->my_count - pipe->count_received;
+            get_idx         = pipe->get_idx % pipe->num_buffers;
+            count      = ucc_min(pipe->buffer_size / dt_size, remaining_elems);
+            get_offset = pipe->count_received * dt_size + pipe->my_offset;
+            data_size  = count * dt_size;
+            src_rank   = pipe->src_rank;
+            getbuf = accbuf->state == FREE ? accbuf : &pipe->getbuf[get_idx];
+            src_addr = PTR_OFFSET(
+                            task->allreduce_sliding_window.bufs->sbufs[src_rank],
+                            get_offset);
+            dst_addr = getbuf->buf;
+
+            ucc_assert(getbuf->state == FREE);
+
+            getbuf->state   = RECVING;
+            getbuf->count   = count;
+            getbuf->bytes   = data_size;
+            ucc_tl_ucp_get_ep(tl_team, src_rank, &ep);
+            getbuf->ucp_req = ucp_get_nbx(
+                ep, dst_addr,
+                data_size, (uint64_t)src_addr,
+                task->allreduce_sliding_window.bufs->src_rkeys[src_rank],
+                &req_param);
+
+            pipe->src_rank = (src_rank + 1) % host_team_size;
+
+            if (getbuf != accbuf) {
+                pipe->avail_buffs--;
+                pipe->get_idx++;
+            }
+
+            pipe->done_get++;
+            if (pipe->done_get == host_team_size) {
+                pipe->count_received += count;
+            }
+        }
+
+        if (accbuf->state == RECVING) {
+            request = accbuf->ucp_req;
+            if (ucc_tl_ucp_allreduce_sliding_window_req_test(request, task) ==
+                UCC_OK) {
+                if (request)
+                    ucp_request_free(request);
+                accbuf->state   = REDUCING;
+                accbuf->ucp_req = NULL;
+            }
```

Comment:
i would recommend to use callbacks for this, you will be able to complete multiple requests in ucp_worker_progress and ucp_req object is actually not needed

Reply:
this is a good optimization, but to meet the july deadline we agreed it would be ok to leave out for now